Time-dependent TPC dEdx calibration

badarots opened the pull request:
A first attempt to write the necessary code to track the MIP position variation over time.

badarots committed:
Initial attempt on TPC dEdx calibration

badarots commented on Detectors/TPC/calibration/CMakeLists.txt:
```diff
@@ -20,9 +20,14 @@ o2_add_library(TPCCalibration
                                      O2::TPCReconstruction ROOT::Minuit
-                                     Microsoft.GSL::GSL)
+                                     Microsoft.GSL::GSL
+                                     O2::DetectorsCalibration
```
I have to check which library I really need to add here, I think a can remove some of those, like O2::DetectorsDCS.

badarots replied:
Removed

wiechula commented on Detectors/TPC/calibration/include/TPCCalibration/CalibdEdx.h:
```diff
@@ -0,0 +1,108 @@
+  ClassDefNV(CalibdEdx, 1);
+};
+
+class CalibratordEdx final : public o2::calibration::TimeSlotCalibration<o2::tpc::TrackTPC, o2::tpc::CalibdEdx>
```
I prefer to put the `Calibrator` in a separate file.

wiechula commented on Detectors/TPC/calibration/include/TPCCalibration/CalibdEdx.h:
```diff
@@ -0,0 +1,108 @@
+  const std::array<Hist, 2>& getHist() const { return mHist; }
+
+ private:
+  int mMinClusters;
+  double mMinP, mMaxP;
```
Default initialize, separate lines and comments:
```suggestion
  int mMinClusters{}; ///< bla
  double mMinP{}; ///< bla
  double mMaxP{}; ///< bla
```

wiechula commented on Detectors/TPC/calibration/include/TPCCalibration/CalibdEdx.h:
```diff
@@ -0,0 +1,108 @@
+  void dumpToFile(const std::string& file_name) const;
+
+ private:
+  int mMinEntries;
+  int mMinClusters;
+  double mMinP, mMaxP;
+  int mNBins;
```
see above, deafault initialize, doxygen comments, one variable per line.

wiechula commented on Detectors/TPC/calibration/include/TPCCalibration/CalibdEdx.h:
```diff
@@ -0,0 +1,108 @@
+  int mMinClusters;
+  double mMinP, mMaxP;
+  int mNBins;
+  CcdbObjectInfoVector mInfoVector; // vector of CCDB Infos , each element is filled with the CCDB description of the accompanying LHCPhase
+  MIPVector mMIPVector;             // vector of LhcPhase, each element is filled in "process" when we finalize one slot (multiple can be finalized during the same "process", which is why we have a vector. Each element is to be considered the output of the device, and will go to the CCDB
```
doxygen comments
```suggestion
  CcdbObjectInfoVector mInfoVector; ///< vector of CCDB Infos , each element is filled with the CCDB description of the accompanying LHCPhase
  MIPVector mMIPVector;             ///< vector of LhcPhase, each element is filled in "process" when we finalize one slot (multiple can be finalized during the same "process", which is why we have a vector. Each element is to be considered the output of the device, and will go to the CCDB
```

wiechula commented on Detectors/TPC/calibration/include/TPCCalibration/FastHisto.h:
```diff
@@ -21,6 +21,7 @@
 #include "Framework/Logger.h"
 #include <sstream>
 #include <vector>
+#include <iomanip>
```
Why is this needed?

wiechula commented on Detectors/TPC/calibration/include/TPCCalibration/CalibdEdx.h:
```diff
@@ -0,0 +1,108 @@
+ private:
+  int mMinClusters;
+  double mMinP, mMaxP;
+  std::array<float, 2> mHistEntries{0, 0};
+  std::array<Hist, 2> mHist;
```
doxygen comments

wiechula commented on Detectors/TPC/calibration/include/TPCCalibration/CalibdEdx.h:
```diff
@@ -0,0 +1,108 @@
+
+// o2 includes
+#include "DataFormatsTPC/TrackTPC.h"
+#include "DataFormatsTPC/dEdxInfo.h"
```
is this needed here?

wiechula commented on Detectors/TPC/workflow/src/CalibdEdxSpec.cxx:
```diff
@@ -0,0 +1,132 @@
+ public:
+  void init(framework::InitContext& ic) final
+  {
+    int slot_length = ic.options().get<int>("tf-per-slot");
+    int max_delay = ic.options().get<int>("max-delay");
+    int min_entries = std::max(50, ic.options().get<int>("min-entries"));
+    double min_p = std::min(0.49, ic.options().get<double>("min-momentum"));
+    double max_p = std::max(0.51, ic.options().get<double>("max-momentum"));
+    int min_clusters = std::max(15, ic.options().get<int>("min-clusters"));
+    int nbins = std::max(50, ic.options().get<int>("nbins"));
```
```suggestion
    const int slot_length = ic.options().get<int>("tf-per-slot");
    const int max_delay = ic.options().get<int>("max-delay");
    const int min_entries = std::max(50, ic.options().get<int>("min-entries"));
    const double min_p = std::min(0.49, ic.options().get<double>("min-momentum"));
    const double max_p = std::max(0.51, ic.options().get<double>("max-momentum"));
    const int min_clusters = std::max(15, ic.options().get<int>("min-clusters"));
    const int nbins = std::max(50, ic.options().get<int>("nbins"));
```

wiechula commented on Detectors/TPC/workflow/src/CalibdEdxSpec.cxx:
```diff
@@ -0,0 +1,132 @@
+
+  void run(ProcessingContext& pc) final
+  {
+    auto tfcounter = o2::header::get<DataProcessingHeader*>(pc.inputs().get("tracks").header)->startTime;
```
```suggestion
   const auto tfcounter = o2::header::get<DataProcessingHeader*>(pc.inputs().get("tracks").header)->startTime;
```

wiechula commented on Detectors/TPC/workflow/src/CalibdEdxSpec.cxx:
```diff
@@ -0,0 +1,132 @@
+    constexpr uint64_t INFINITE_TF = 0xffffffffffffffff;
+    mCalibrator->checkSlotsToFinalize(INFINITE_TF);
+    sendOutput(eos.outputs());
+    mCalibrator->dumpToFile("mip_position.root");
```
Please add an option that this is only dumped if activated in the workflow, like here https://github.com/AliceO2Group/AliceO2/blob/dev/Detectors/TPC/workflow/include/TPCWorkflow/TPCCalibPedestalSpec.h#L180

wiechula commented on Detectors/TPC/calibration/include/TPCCalibration/CalibdEdx.h:
```diff
@@ -0,0 +1,108 @@
+#include "DetectorsCalibration/TimeSlot.h"
+#include "TPCCalibration/FastHisto.h"
+
+namespace o2
+{
+namespace tpc
+{
```
please change this everywhere
```suggestion
namespace o2::tpc
{
```

wiechula commented on Detectors/TPC/workflow/src/CalibdEdxSpec.cxx:
```diff
@@ -0,0 +1,132 @@
+      {"tf-per-slot", VariantType::Int, 100, {"number of TFs per calibration time slot"}},
+      {"max-delay", VariantType::Int, 3, {"number of slots in past to consider"}},
+      {"min-entries", VariantType::Int, 100, {"minimum number of entries to fit single time slot"}},
+      {"min-momentum", VariantType::Double, 0.4, {"minimum momentum cut"}},
+      {"max-momentum", VariantType::Double, 0.6, {"maximum momentum cut"}},
+      {"min-clusters", VariantType::Int, 60, {"minimum number of clusters in a track"}},
```
at some point we should put this into a track cut class. For the moment it is fine.

wiechula reviewed (dismissed):
I think for the moment this is fine. Please mark ready for review. We'll continue then with further developments in another PR.

wiechula commented on Detectors/TPC/calibration/include/TPCCalibration/dEdxHistos.h:
```diff
@@ -0,0 +1,82 @@
+  dEdxHistos() = default;
+
+  /// Constructor that enable tracks cuts
+  dEdxHistos(TrackCuts cuts, unsigned int nBins);
```
```suggestion
  dEdxHistos(const TrackCuts& cuts, unsigned int nBins);
```
below you move, which will 'return' an empty input object. Not necessarily what one would expect I think.

wiechula commented on Detectors/TPC/calibration/include/TPCCalibration/dEdxHistos.h:
```diff
@@ -0,0 +1,82 @@
+  const std::array<Hist, 2>& getHist() const { return mHist; }
+
+  /// Save the histograms to a file
+  void dumpToFile(const std::string& fileName) const;
```
```suggestion
  void dumpToFile(std::string_view fileName) const;
```

wiechula commented on Detectors/TPC/calibration/include/TPCCalibration/dEdxHistos.h:
```diff
@@ -0,0 +1,82 @@
+  dEdxHistos(double minP, double maxP, int minClusters, unsigned int nBins);
+
+  /// Constructor that disables any tracks cuts
+  dEdxHistos(unsigned int nBins);
```
Perhaps it is better to set default Track cuts and steer if they should be applied by the flag `mApplyCuts` with setters and getter.

badarots replied:
You are right, this added constructor just add complexity.

wiechula commented on Detectors/TPC/calibration/src/CalibdEdx.cxx:
```diff
@@ -12,73 +12,101 @@
+
+  mDumpTree->Branch("timeFrame", &mTimeFrameBranch, "timeFrame/l");
+  mDumpTree->Branch("calibMIP", mMIPBranch.get());
+  mDumpTree->Branch("dEdxHistos", mHistosBranch.get());
```
It might be simpler to use the `TreeStream` functionality. Then you don't need to keep branches for data memebers that should be stored in the tree. See e.g. how it is done here:
https://github.com/AliceO2Group/AliceO2/blob/dev/Detectors/TPC/calibration/include/TPCCalibration/CalibLaserTracks.h#L120
https://github.com/AliceO2Group/AliceO2/blob/dev/Detectors/TPC/calibration/src/CalibLaserTracks.cxx#L82

badarots replied:
This is nice! I really struggled to get this data-keeping thing right, and it is so ugly. I wasn't aware of another way to do it and I am really glad you point this out. I will changes this right away!

wiechula commented on Detectors/TPC/calibration/src/dEdxHistos.cxx:
```diff
@@ -0,0 +1,86 @@
+
+using namespace o2::tpc;
+
+dEdxHistos::dEdxHistos(TrackCuts cuts, unsigned int nBins)
```
see comment above

wiechula commented on Detectors/TPC/calibration/src/dEdxHistos.cxx:
```diff
@@ -0,0 +1,86 @@
+{
+  for (const auto& track : tracks) {
+    const auto p = track.getP();
+    const int cluster_count = track.getNClusters();
```
Naming convention camel case!

badarots replied:
It is stronger than me!

wiechula commented on Detectors/TPC/workflow/src/MIPTrackFilterSpec.cxx:
```diff
@@ -0,0 +1,96 @@
+{
+  const auto tracks = pc.inputs().get<gsl::span<TrackTPC>>("tracks");
+
+  std::copy_if(tracks.begin(), tracks.end(), std::back_inserter(mMIPTracks),
+               [this](const auto& track) { return this->mCuts.goodTrack(track); });
```
cool, I like this one!

wiechula reviewed (changes requested):
Thanks for the updates! It would be good if you could squash everything, since it is hard to read the diffs now after classes were renamed. You could keep 2 commits only:
1. moving of TrackCuts and all related changes
2. all MIP calibration stuff

Afterwards I'll go through carefully once more and have a look.

For the track cuts it might be nicer to have this configurable `ParameterTrackCuts` using a `ConfigurableParam` as e.g. here: https://github.com/AliceO2Group/AliceO2/blob/dev/Detectors/TPC/base/include/TPCBase/ParameterDetector.h and add function like `setCuts(TrackCuts& cuts)`. Then one can configure the cuts easily from the command line. We can discuss details.

badarots committed:
Move TrackCuts to DataFormatsTPC

badarots commented:
I reworked the commit contents as you asked and applied all requested changes, but the `ConfigurableParam` stuff.

wiechula commented on Detectors/TPC/calibration/include/TPCCalibration/CalibdEdx.h:
```diff
@@ -0,0 +1,113 @@
+  Slot& emplaceNewSlot(bool front, TFType tstart, TFType tend) final;
+
+  void setApplyCuts(bool apply) { mApplyCuts = apply; }
+  bool getApplyCuts() { return mApplyCuts; }
```
```suggestion
  bool getApplyCuts() const { return mApplyCuts; }
```

wiechula commented on Detectors/TPC/calibration/src/CalibdEdx.cxx:
```diff
@@ -0,0 +1,104 @@
+  return slot;
+}
+
+void CalibdEdx::enableDumpToFile(std::string_view fileName)
```
perhaps `enableDebugStream` or `enableDebugOutput`.

wiechula commented on Detectors/TPC/calibration/src/CalibdEdx.cxx:
```diff
@@ -0,0 +1,104 @@
+  mDumpStream = std::make_unique<o2::utils::TreeStreamRedirector>(fileName.data(), "recreate");
+}
+
+void CalibdEdx::disableDumpToFile()
```
Same as above.

wiechula commented on Detectors/TPC/calibration/src/CalibdEdx.cxx:
```diff
@@ -0,0 +1,104 @@
+  mDumpStream.reset();
+}
+
+void CalibdEdx::dumpToFile() const
```
Also here the name is a bit misleading, since the dumping to file using the stream functionality is done in the place where it is used. Usually, dumpToFile should really put contents to a file. But this is not what is done here. So one might call it `finalizeDebugOutput` or so.

wiechula commented on Detectors/TPC/calibration/include/TPCCalibration/dEdxHistos.h:
```diff
@@ -0,0 +1,85 @@
+  double getCSideEntries() const { return mEntries[1]; }
+
+  /// Get the underlying histograms
+  const std::array<Hist, 2>& getHist() const { return mHist; }
```
perhaps
```suggestion
  const std::array<Hist, 2>& getHists() const { return mHist; }
```
since it is not only one.

wiechula reviewed (changes requested):
Thanks a lot @badarots! Only minor things left. Please amend them to the latest commit.

badarots commented:
Hi @wiechula, I applied the requested changes! As you can see I changed the `CalibMIP` from a `std::array` to a custom struct, and added it to the root dictionary. Now the debug output to file is working correctly, but this needs to be changed in the future.

wiechula reviewed (dismissed):
@badarots , thanks for all the changes.

badarots commented:
What now @wiechula? Should I squash the last two commits and proceed to the merge of this PR or do you prefer that I finish the `ParameterTrackCuts` first?

wiechula commented:
@badarots , merging can only be done by the code owners. Yes, please squash the last two commits. I'll then approve once more and after all test have finished successfully I'll ask to have it merged.

wiechula commented:
And make it ready for review, of course.

shahor02 reviewed (changes requested):
@badarots the o2 CI failure is genuine:
````
/mnt/mesos/sandbox/sandbox/o2/sw/SOURCES/O2/6305/0/Detectors/TPC/calibration/src/dEdxHistos.cxx: In member function 'void o2::tpc::dEdxHistos::merge(const o2::tpc::dEdxHistos*)':
/mnt/mesos/sandbox/sandbox/o2/sw/SOURCES/O2/6305/0/Detectors/TPC/calibration/src/dEdxHistos.cxx:57:24: error: 'prev' was not declared in this scope; did you mean 'std::prev'?
   57 |     assert(binCount == prev->getHist()[i].getNBins());
      |                        ^~~~
````

wiechula commented on Detectors/TPC/calibration/src/dEdxHistos.cxx:
```diff
@@ -0,0 +1,77 @@
+{
+  for (size_t i = 0; i < mHist.size(); i++) {
+    const auto binCount = mHist[i].getNBins();
+    assert(binCount == prev->getHist()[i].getNBins());
```
I guess this should be
```suggestion
    assert(binCount == other->getHist()[i].getNBins());
```
@badarots , can you please check.

badarots commented:
As discussed with @wiechula today, we agreed to just remove that assertion line.

badarots commented:
So there is an error in the `ensure-executable-naming-convention` test
```
ERROR: executable name o2-tpc-calib-dEdx does not follow naming convention : should only contains lowercase letters, numbers, or dashes
```

Should I just change the executable name to `o2-tpc-calib-dedx` (lowercase "de")?

shahor02 commented on Detectors/TPC/workflow/CMakeLists.txt:
```diff
@@ -103,6 +105,11 @@ o2_add_executable(idc-to-vector
                   SOURCES src/tpc-idc-to-vector.cxx
                   PUBLIC_LINK_LIBRARIES O2::TPCWorkflow)
 
+o2_add_executable(calib-dEdx
```
yes, just change this to `calib-dedx`

wiechula commented:
@shahor02 , @davidrohr , I think the macOS failure is unrelated.

davidrohr commented:
Probably, I have seen such errors often on the MacCI recently.
@TimoWilken @MichaelLettrich : Could you check why the MacCI fails here?

shahor02 commented:
I thought this problem of compiled macro's test was fixed already... Anyway, the macOS log also contains legitimate warnings:
````
/System/Volumes/Data/build/ci_checks/alice_ci_0/o2/sw/SOURCES/O2/6305/0/Detectors/TPC/qc/macro/runPID.C:55:23: warning: comparison of integers of different signs: 'int' and 'size_t' (aka 'unsigned long') [-Wsign-compare]
    for (int k = 0; k < nTracks; k++) {
````
````
/System/Volumes/Data/build/ci_checks/alice_ci_0/o2/sw/SOURCES/O2/6305/0/Detectors/TPC/qc/macro/runTracks.C:59:23: warning: comparison of integers of different signs: 'int' and 'size_t' (aka 'unsigned long') [-Wsign-compare]
    for (int k = 0; k < nTracks; k++) {
                    ~ ^ ~~~~~~~
````

wiechula commented:
Indeed, but these lines were not touched by this commit.
@badarots , can you fix it nevertheless in both places:
```c++
for (size_t k = 0; k < nTracks; k++) {
```
should do.

badarots commented:
I changed `int` -> `size_t` as requested. I'm not sure if I was supposed to amend my last commit or create a new one for this little change. So I amended it.

wiechula commented:
@shahor02 , can this be merged? I think the macOS failure is again unrelated. cs8 fails with `No space left on device`.
